File: skills/sangkayana/tally.py
#!/usr/bin/env python3
"""Tally skill usage evidence for the sangkayana council.

Prints an evidence horizon, then one row per installed skill:
  last_used  typed  model  installed  age_days  invoke  desc_words  class  name

class:  managed (lock-tracked symlink) | foreign (hand-placed dir) | external (symlink elsewhere)
invoke: model (description loaded every turn) | user (typed only, zero context load)
desc_words: size of the always-loaded description; the context load a removal reclaims.
last_used: date of most recent evidence, or "-" when the record is silent.
"""
import json, os, re, glob, datetime, sys

H = os.path.expanduser("~")
SKILLS = f"{H}/.claude/skills"
LOCK = f"{H}/.agents/.skill-lock.json"
HIST = f"{H}/.claude/history.jsonl"
TRANSCRIPTS = f"{H}/.claude/projects/*/*.jsonl"

day = lambda ms: datetime.date.fromtimestamp(ms / 1000).isoformat()
today = datetime.date.today()


def load_lock():
    try:
        return json.load(open(LOCK))["skills"]
    except Exception:
        return {}


def classify(name):
    p = f"{SKILLS}/{name}"
    if not os.path.islink(p):
        return "foreign"
    target = os.path.realpath(p)
    return "managed" if f"{os.sep}.agents{os.sep}skills{os.sep}" in target + os.sep else "external"


def frontmatter(name):
    """(invoke, desc_words) from the skill's frontmatter."""
    for path in (f"{SKILLS}/{name}/SKILL.md", f"{SKILLS}/{name}/skill.md"):
        try:
            head = open(path, errors="ignore").read(8000)
        except OSError:
            continue
        m = re.match(r"\s*---\s*\n(.*?)\n---", head, re.S)
        if not m:
            return "?", 0
        fm = m.group(1)
        user_only = re.search(r"^disable-model-invocation:\s*true", fm, re.M | re.I)
        d = re.search(r"^description:\s*(.*(?:\n[ \t]+.*)*)", fm, re.M)
        words = len((d.group(1) if d else "").split())
        return ("user" if user_only else "model"), words
    return "?", 0


def typed_usage():
    """Typed /invocations from history.jsonl. Complete back to the file's first entry."""
    hits, stamps = {}, []
    try:
        lines = open(HIST, errors="ignore").readlines()
    except OSError:
        return {}, None
    for line in lines:
        try:
            d = json.loads(line)
        except ValueError:
            continue
        ts = d.get("timestamp")
        if ts:
            stamps.append(ts)
        m = re.match(r"\s*/([a-zA-Z0-9_-]+)", d.get("display") or "")
        if m and ts:
            n = m.group(1)
            c, last = hits.get(n, (0, 0))
            hits[n] = (c + 1, max(last, ts))
    return hits, (min(stamps) if stamps else None)


def model_usage():
    """Skill tool-calls in surviving session transcripts. Older sessions get swept."""
    hits, stamps = {}, []
    for f in glob.glob(TRANSCRIPTS):
        try:
            body = open(f, errors="ignore").read()
            mtime = int(os.path.getmtime(f) * 1000)
        except OSError:
            continue
        stamps.append(mtime)
        for m in re.finditer(r'"skill"\s*:\s*"([a-zA-Z0-9_:-]+)"', body):
            n = m.group(1).split(":")[-1]
            c, last = hits.get(n, (0, 0))
            hits[n] = (c + 1, max(last, mtime))
    return hits, (min(stamps) if stamps else None)


def main():
    lock = load_lock()
    try:
        names = sorted(n for n in os.listdir(SKILLS) if not n.startswith("."))
    except OSError:
        sys.exit(f"no skills directory at {SKILLS}")

    typed, typed_from = typed_usage()
    model, model_from = model_usage()

    print("EVIDENCE HORIZON")
    print(f"  typed  (history.jsonl)  reaches back to {day(typed_from) if typed_from else 'nothing recorded'}")
    print(f"  model  (transcripts)    reaches back to {day(model_from) if model_from else 'nothing recorded'}")
    print("  Silence before a horizon is missing evidence, not disuse.")
    print(f"  Skills shared with other agents may have fired outside {SKILLS} entirely.\n")

    rows = []
    for n in names:
        tc, tl = typed.get(n, (0, 0))
        mc, ml = model.get(n, (0, 0))
        last = max(tl, ml)
        inst = (lock.get(n) or {}).get("installedAt", "")
        try:
            age = (today - datetime.date.fromisoformat(inst[:10])).days
        except ValueError:
            inst, age = "unknown", ""
        inv, dw = frontmatter(n)
        rows.append((last, tc + mc, {
            "last_used": day(last) if last else "-",
            "typed": tc, "model": mc,
            "installed": inst[:10] if inst else "unknown",
            "age_days": age, "invoke": inv, "desc_words": dw,
            "class": classify(n), "name": n,
        }))

    rows.sort(key=lambda r: (r[0], r[1]))
    hdr = ("last_used", "typed", "model", "installed", "age_days",
           "invoke", "desc_words", "class", "name")
    print("\t".join(hdr))
    for _, _, r in rows:
        print("\t".join(str(r[k]) for k in hdr))

    silent = sum(1 for _, t, _ in rows if t == 0)
    load = sum(r["desc_words"] for _, _, r in rows if r["invoke"] == "model")
    quiet_load = sum(r["desc_words"] for _, t, r in rows
                     if r["invoke"] == "model" and t == 0)
    noun = "skill" if len(rows) == 1 else "skills"
    print(f"\n{len(rows)} {noun} in the canon; {silent} with no surviving evidence of use.")
    print(f"context load: ~{load} description words loaded every turn, "
          f"~{quiet_load} of them from skills with no recorded use.")


if __name__ == "__main__":
    main()

File: skills/in-progress/sangkayana/tally.ts
#!/usr/bin/env node
/**
 * Tally skill usage evidence for the sangkayana council.
 *
 * Prints an evidence horizon, then one row per installed skill:
 *   last_used  typed  model  installed  age_days  invoke  desc_words  class  name
 *
 * class:  managed (lock-tracked symlink) | foreign (hand-placed dir) | external (symlink elsewhere)
 * invoke: model (description loaded every turn) | user (typed only, zero context load)
 * desc_words: size of the always-loaded description; the context load a removal reclaims.
 * last_used: date of most recent evidence, or "-" when the record is silent.
 *
 * Runs on Node 22.18+ with no dependencies and no build step.
 */
import { globSync, lstatSync, readFileSync, readdirSync, realpathSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { join, sep } from "node:path";

const HOME = homedir();
const SKILLS = join(HOME, ".claude", "skills");
const LOCK = join(HOME, ".agents", ".skill-lock.json");
const HISTORY = join(HOME, ".claude", "history.jsonl");
const TRANSCRIPTS = join(HOME, ".claude", "projects", "*", "*.jsonl");

type Invoke = "model" | "user" | "?";
type Provenance = "managed" | "foreign" | "external";

/** One skill's usage in a single evidence stream. */
interface Usage {
  count: number;
  /** epoch ms of the most recent hit */
  last: number;
}

interface Row {
  /** epoch ms of the most recent evidence; sorts within a day, never printed */
  lastMs: number;
  last_used: string;
  typed: number;
  model: number;
  installed: string;
  age_days: number | "";
  invoke: Invoke;
  desc_words: number;
  class: Provenance;
  name: string;
}

const COLUMNS = [
  "last_used", "typed", "model", "installed", "age_days",
  "invoke", "desc_words", "class", "name",
] as const satisfies readonly (keyof Row)[];

/** Local calendar date of an epoch-ms stamp, as YYYY-MM-DD. */
function day(ms: number): string {
  const d = new Date(ms);
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`;
}

function read(path: string): string | null {
  try {
    return readFileSync(path, "utf8");
  } catch {
    return null;
  }
}

function bump(into: Map<string, Usage>, name: string, at: number): void {
  const prev = into.get(name);
  into.set(name, { count: (prev?.count ?? 0) + 1, last: Math.max(prev?.last ?? 0, at) });
}

function loadLock(): Record<string, { installedAt?: string }> {
  const raw = read(LOCK);
  if (!raw) return {};
  try {
    return JSON.parse(raw).skills ?? {};
  } catch {
    return {};
  }
}

function provenance(name: string): Provenance {
  const path = join(SKILLS, name);
  try {
    if (!lstatSync(path).isSymbolicLink()) return "foreign";
    const target = realpathSync(path) + sep;
    return target.includes(`${sep}.agents${sep}skills${sep}`) ? "managed" : "external";
  } catch {
    return "foreign";
  }
}

/** Invocation mode and description size, from the skill's frontmatter. */
function frontmatter(name: string): Pick<Row, "invoke" | "desc_words"> {
  for (const file of ["SKILL.md", "skill.md"]) {
    const body = read(join(SKILLS, name, file));
    if (body === null) continue;
    const matched = /^\s*---\s*\n([\s\S]*?)\n---/.exec(body);
    if (!matched) return { invoke: "?", desc_words: 0 };
    const [, fm = ""] = matched;
    const userOnly = /^disable-model-invocation:\s*true/im.test(fm);
    const [, desc = ""] = /^description:[ \t]*(.*(?:\n[ \t]+.*)*)/m.exec(fm) ?? [];
    const words = desc.split(/\s+/).filter(Boolean).length;
    return { invoke: userOnly ? "user" : "model", desc_words: words };
  }
  return { invoke: "?", desc_words: 0 };
}

/** Typed /invocations. Complete back to the first entry in history.jsonl. */
function typedUsage(): { hits: Map<string, Usage>; from: number | null } {
  const hits = new Map<string, Usage>();
  const body = read(HISTORY);
  if (body === null) return { hits, from: null };
  let from: number | null = null;
  for (const line of body.split("\n")) {
    if (!line.trim()) continue;
    let entry: { display?: string; timestamp?: number };
    try {
      entry = JSON.parse(line);
    } catch {
      continue;
    }
    const at = entry.timestamp;
    if (typeof at !== "number") continue;
    from = from === null ? at : Math.min(from, at);
    const [, invoked] = /^\s*\/([a-zA-Z0-9_-]+)/.exec(entry.display ?? "") ?? [];
    if (invoked) bump(hits, invoked, at);
  }
  return { hits, from };
}

/** Skill tool-calls in surviving session transcripts. Older sessions get swept. */
function modelUsage(): { hits: Map<string, Usage>; from: number | null } {
  const hits = new Map<string, Usage>();
  let from: number | null = null;
  for (const file of globSync(TRANSCRIPTS)) {
    const body = read(file);
    if (body === null) continue;
    let mtime: number;
    try {
      mtime = statSync(file).mtimeMs;
    } catch {
      continue;
    }
    from = from === null ? mtime : Math.min(from, mtime);
    for (const [, invoked] of body.matchAll(/"skill"\s*:\s*"([a-zA-Z0-9_:-]+)"/g)) {
      // plugin skills arrive as "plugin:skill"; the council judges the skill
      if (invoked) bump(hits, invoked.split(":").at(-1) ?? invoked, mtime);
    }
  }
  return { hits, from };
}

function main(): void {
  const lock = loadLock();
  let names: string[];
  try {
    names = readdirSync(SKILLS).filter((n) => !n.startsWith(".")).sort();
  } catch {
    console.error(`no skills directory at ${SKILLS}`);
    process.exit(1);
  }

  const typed = typedUsage();
  const model = modelUsage();

  console.log("EVIDENCE HORIZON");
  console.log(`  typed  (history.jsonl)  reaches back to ${typed.from ? day(typed.from) : "nothing recorded"}`);
  console.log(`  model  (transcripts)    reaches back to ${model.from ? day(model.from) : "nothing recorded"}`);
  console.log("  Silence before a horizon is missing evidence, not disuse.");
  console.log(`  Skills shared with other agents may have fired outside ${SKILLS} entirely.\n`);

  const today = new Date();
  const rows: Row[] = names.map((name) => {
    const t = typed.hits.get(name);
    const m = model.hits.get(name);
    const last = Math.max(t?.last ?? 0, m?.last ?? 0);
    const installedAt = lock[name]?.installedAt?.slice(0, 10);
    const installed = new Date(`${installedAt}T00:00:00`);
    const known = installedAt !== undefined && !Number.isNaN(installed.valueOf());
    return {
      lastMs: last,
      last_used: last ? day(last) : "-",
      typed: t?.count ?? 0,
      model: m?.count ?? 0,
      installed: known ? installedAt! : "unknown",
      age_days: known ? Math.floor((today.valueOf() - installed.valueOf()) / 86_400_000) : "",
      ...frontmatter(name),
      class: provenance(name),
      name,
    };
  });

  rows.sort((a, b) =>
    a.lastMs - b.lastMs ||
    (a.typed + a.model) - (b.typed + b.model) ||
    a.name.localeCompare(b.name));

  console.log(COLUMNS.join("\t"));
  for (const row of rows) console.log(COLUMNS.map((c) => row[c]).join("\t"));

  const silent = rows.filter((r) => r.typed + r.model === 0);
  const load = (rs: Row[]) =>
    rs.filter((r) => r.invoke === "model").reduce((sum, r) => sum + r.desc_words, 0);
  console.log(`\n${rows.length} ${rows.length === 1 ? "skill" : "skills"} in the canon; ` +
    `${silent.length} with no surviving evidence of use.`);
  console.log(`context load: ~${load(rows)} description words loaded every turn, ` +
    `~${load(silent)} of them from skills with no recorded use.`);
}

main();
